File: app/src/main/java/com/hover/stax/channels/ChannelDao.java
package com.hover.stax.channels;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Update;

import java.util.List;

@Dao
public interface ChannelDao {
	@Query("SELECT * FROM channels ORDER BY country_alpha2, name")
	LiveData<List<Channel>> getAll();

	@Query("SELECT * FROM channels WHERE selected = :selected ORDER BY defaultAccount DESC")
	LiveData<List<Channel>> getSelected(boolean selected);

	@Query("SELECT * FROM channels WHERE defaultAccount = 'true' ORDER BY defaultAccount DESC")
	LiveData<Channel> getDefault();

	@Query("SELECT * FROM channels WHERE country_alpha2 = :countryAlpha2")
	LiveData<List<Channel>> getByCountry(String countryAlpha2);

//	@Query("SELECT * FROM channels WHERE hni_list IN :hniList")
//	MutableLiveData<List<Channel>> getByHniList(String[] hniList);

	@Query("SELECT EXISTS(SELECT id FROM channels WHERE id = :id LIMIT 1)")
	Boolean exists(int id);

	@Query("SELECT * FROM channels WHERE id = :id LIMIT 1")
	Channel getChannel(int id);

	@Query("SELECT * FROM channels WHERE id = :id LIMIT 1")
	LiveData<Channel> getLiveChannel(int id);

	@Insert(onConflict = OnConflictStrategy.IGNORE)
	void insertAll(Channel... channels);

	@Insert(onConflict = OnConflictStrategy.IGNORE)
	void insert(Channel channel);

	@Update
	void update(Channel channel);

	@Delete
	void delete(Channel channel);

	@Query("DELETE FROM channels")
	void deleteAll();
}

File: app/src/main/java/com/hover/stax/data/local/channels/ChannelDao.kt
package com.hover.stax.data.local.channels

import androidx.lifecycle.LiveData
import androidx.room.*
import com.hover.stax.accounts.ChannelWithAccounts
import com.hover.stax.domain.model.Channel
import kotlinx.coroutines.flow.Flow

@Dao
interface ChannelDao {

    @get:Query("SELECT * FROM channels WHERE published = 1 ORDER BY isFavorite DESC, name ASC")
    val publishedChannels: Flow<List<Channel>>

    @get:Query("SELECT * FROM channels ORDER BY name ASC")
    val allChannels: LiveData<List<Channel>>

    @Query("SELECT * FROM channels WHERE selected = :selected ORDER BY defaultAccount DESC, name ASC")
    fun getSelected(selected: Boolean): LiveData<List<Channel>>

    @Query("SELECT * FROM channels WHERE id IN (:channel_ids) ORDER BY name ASC")
    suspend fun getChannelsByIds(channel_ids: List<Int>): List<Channel>

    @Query("SELECT * FROM channels WHERE id IN (:channel_ids) ORDER BY name ASC")
    fun getChannels(channel_ids: IntArray): LiveData<List<Channel>>

    @Query("SELECT * FROM channels WHERE country_alpha2 = :countryCode ORDER BY name ASC")
    suspend fun getChannels(countryCode: String): List<Channel>

    @Query("SELECT * FROM channels WHERE country_alpha2 in (:countryCodes) ORDER BY name ASC")
    suspend fun getChannels(countryCodes: List<String>): List<Channel>

    @Query("SELECT * FROM channels WHERE country_alpha2 = :countryCode AND id IN (:channel_ids) ORDER BY name ASC")
    suspend fun getChannels(countryCode: String, channel_ids: IntArray): List<Channel>

    @Query("SELECT * FROM channels WHERE id = :id LIMIT 1")
    suspend fun getChannel(id: Int): Channel?

    @Query("SELECT * FROM channels WHERE id = :id LIMIT 1")
    fun getLiveChannel(id: Int): LiveData<Channel>

    @get:Query("SELECT * FROM channels")
    val channels: List<Channel>

    @Transaction
    @Query("SELECT * FROM channels where selected = 1 ORDER BY name ASC")
    fun getChannelsAndAccounts(): List<ChannelWithAccounts>

    @Transaction
    @Query("SELECT * FROM channels where id = :id ORDER BY name ASC")
    fun getChannelAndAccounts(id: Int): ChannelWithAccounts?

    @get:Query("SELECT COUNT(id) FROM channels")
    val dataCount: Int

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    suspend fun insertAll(vararg channels: Channel?)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insert(channel: Channel?)

    @Update
    suspend fun update(channel: Channel)

    @Update
    suspend fun updateAll(channel: List<Channel>)

    @Delete
    suspend fun delete(channel: Channel)

    @Query("DELETE FROM channels")
    suspend fun deleteAll()

}